### ISSUE TITLE
Set environment variables

### ISSUE THREAD

**Issue by jimmykarily** [epinio/epinio#375]
Depends on #347 

Various applications may need to be configured through environment variables. For example, a rails application needs to have the `RAILS_MASTER_KEY` environment variable set to a specific secret (https://edgeguides.rubyonrails.org/security.html#custom-credentials).

## Acceptance Criteria

The user can specify env vars for a created app (i.e. when creating the app). See https://github.com/epinio/epinio/issues/375#issuecomment-868301329

When pushing the app, staging will create a deployment. The deployment will mount the secret as env vars.

mm: I'd suggest to postpone `--staging` env vars for a separate follow-up story.

## Previous Notes

Also buildpacks are moving away from `buildpack.yaml` configuration towards env vars (https://github.com/paketo-buildpacks/nginx/blob/main/rfcs/0001-buildpack-yml-to-env-var.md).

We need a way to allow the user to set environment variables that will be available to the application at run time.

We should also investigate if the same solution would work for staging with buildpacks (see the link above).

## Previous Questions

- Where are the variables stored? Is it on the deployment itself?
- What happens when we push the application again?
- How does the user delete a variable?
- Can the solution look similar to the concept of services? E.g. each environment variable being a secret behind the scenes that can be mounted to any application, kinda like a service, so it has its own lifecycle.

**Comment by jimmykarily** [comment]
The buildpack configuration env vars probably have to be defined on the build container: https://paketo.io/docs/buildpacks/language-family-buildpacks/java/#building-from-source

**Comment by agracey** [comment]
It might also be nice to allow setting a whole config map or secret worth of variables with a yaml file

**Comment by manno** [comment]
The user can still use the [`project.toml`](https://github.com/buildpacks/spec/blob/main/extensions/project-descriptor.md) to replace `buildpack.yml`.

**Comment by jimmykarily** [comment]
Some useful findings on the comments and linked branch here: https://github.com/epinio/epinio/issues/441

**Comment by jimmykarily** [comment]
Extracted the bump to 0.3 of the buildpacks pipeline to a PR: https://github.com/epinio/epinio/pull/453/files

**Comment by jimmykarily** [comment]
This is probably the way to go to solve this issue: https://github.com/epinio/epinio/issues/347

**Comment by jimmykarily** [comment]
We decided to introduce the app crd first (see #347) so we have a resource to which to "attach" the environment variables to. The environment variables will simply be Kubernetes secrets. Probably one for staging environment variables and one for runtime, or one secret for both e.g.

```
env:
  - name: RAILS_MASTER_KEY
    value: 1234
    staging: true
    runtime: false
```

**Comment by jimmykarily** [comment]
TODO:
- env var CRUD (cli and API endpoints):
  - `epinio env var set [--staging] myapp MYVAR value`
  - `epinio env var list [--staging] myapp`
  - `epinio env var delete [--staging] myapp MYVAR`
- combos:
  `epinio app push sample -e MYVAR=value`

**Comment by jimmykarily** [comment]
Working branch: https://github.com/epinio/epinio/tree/375-env-vars

**Comment by kkaempf** [comment]
Why the `env var` (two args) instead of `env` (or `envvar`) ?
Which other `env`-related tasks are there besides variable management ?

**Comment by jimmykarily** [comment]
sorry `env` was supposed to be `app`. I did the first wrong and the copied-pasted all of them. Fixing.

**Comment by andreas-kupries** [comment]
Working PR #587 ...

**Comment by andreas-kupries** [comment]
Working on testing if my work to inject the initial env when pushing a workload for a previously created app I find that a number of `app` command do not work for an app without workload. Namely `list`, `show`, and all of the new `app env` commands. As they all use the same kind of app validation. They all believe that the app does not exist.

Verified the behaviour on `main` branch, commit c7d14e711666f75d5eb4cd3613c3c17f6708733e.
See session below.
I will now have to fix this as I am unable to set env variables on an app without workload to test with, without that.

```
work@tagetarl:~/SUSE/dev/epinio> ep app create hello

🚢 Create application
Organization: workspace
Application: hello

✔️  Ok

work@tagetarl:~/SUSE/dev/epinio> ep app list ; ep app show hello

🚢 Listing applications
Organization: workspace

✔️  Epinio Applications:
| NAME | STATUS | ROUTES | SERVICES |
|------|--------|--------|----------|

🚢 Show application details
Organization: workspace
Application: hello
error showing app: Not Found: {"errors":[{"status":404,"title":"Application 'hello' does not exist","details":""}]}

work@tagetarl:~/SUSE/dev/epinio> ep version

Epinio Version: v0.0.18-9-gc7d14e7
Go Version: go1.14.15

work@tagetarl:~/SUSE/dev/epinio> gl | head -5

commit c7d14e711666f75d5eb4cd3613c3c17f6708733e
Merge: 5ce7624 7b9ac72
Author: Andreas Kupries <andreas.kupries@suse.com>
Date:   Mon Jul 5 13:30:15 2021 +0200
```

**Comment by andreas-kupries** [comment]
Most of the code seems to use `application.Lookup()`, which expects a workload.
I suspect the code should use `application.Exists()`.
Except for listing the apps, that needs some other changes.

**Comment by andreas-kupries** [comment]
Looking deeper ... Lookup delivers a fuller app struct, exists only comes with a bool. For something like `app show` this is not enough ... Hm. That one might need both ... If it exists, but lookup fails, then we fake an app structure, with most things simply empty.

**Comment by manno** [comment]
What I saw is that a lot of the code doesn't actually need a rich app model, the AppRef (name,org) is often enough. You then know an app exists and how to get it.
`app show` needs more information, but I would opt for specific structs, instead of one big app model. `app show` is a good example, because it basically expects a "view model", something that can be more or less directly rendered into a table for CLI output.
More specific models also keep the number of trips to the k8s API server low. I.e. we don't need to fetch workloads (deployments) or the app CRD, unless the code needs data from them.
